### PR TITLE
Add is_countable under PHP 7.3

### PR DESCRIPTION
**Pull request by carusogabriel**
As requested by @nicolas-grekas :)

Waiting for https://github.com/php/php-src/pull/3026

### CODE REVIEW

File: src/Php73/bootstrap.php [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+use Symfony\Polyfill\Php73 as p;
+
+if (PHP_VERSION_ID < 70300) {
+    if (!function_exists('is_countable')) {
+        function is_countable($s) { return p\Php73::is_countable($s); }
```
Comment (nicolas-grekas) [inline src/Php73/bootstrap.php]: for such a simple implementation, I think we can inline the code a remove the static class (as done on other polyfills AFAIK)

File: tests/Php73/Php72Test.php [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+namespace Symfony\Polyfill\Tests\Php73;
+
+use PHPUnit\Framework\TestCase;
+use Symfony\Polyfill\Php73\Php73 as p;
+
+/**
+ * @author Gabriel Caruso <carusogabriel34@gmail.com>
+ *
+ * @covers Symfony\Polyfill\Php73\Php73::<!public>
```
Comment (nicolas-grekas) [inline tests/Php73/Php72Test.php]: should be removed in you accept the previous comment (same below)

File: src/Php73/bootstrap.php (hunk 1 above)
Comment (carusogabriel) [reply src/Php73/bootstrap.php]: @nicolas-grekas Sure, gonna do this!

File: tests/Php73/Php72Test.php [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+{
+    /**
+     * @covers Symfony\Polyfill\Php73\Php73::is_countable
+     */
+    public function testIsCountable()
+    {
+        $this->assertTrue(is_countable([1, 2, 3]));
+        $this->assertTrue(is_countable(new \ArrayIterator(['foo', 'bar', 'baz'])));
+        $this->assertTrue(is_countable(new \ArrayIterator()));
+        $this->assertFalse(is_countable(new \stdClass()));
```
Comment (keradus) [inline tests/Php73/Php72Test.php]: please add a case with generator

File: src/Php73/composer.json [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        "php": ">=5.3.3"
+    },
+    "autoload": {
+        "psr-4": { "Symfony\\Polyfill\\Php73\\": "" },
+        "files": [ "bootstrap.php" ]
+    },
+    "minimum-stability": "dev",
+    "extra": {
+        "branch-alias": {
+            "dev-master": "1.7-dev"
```
Comment (stof) [inline src/Php73/composer.json]: should be 1.8 (and we should bump the version in master before merging this), as this is a new feature and so will be a new minor version, not a new patch version

File: tests/Php73/Php72Test.php (hunk 3 above)
Comment (carusogabriel) [reply tests/Php73/Php72Test.php]: @keradus Is this possible with 5.x tests?

File: src/Php73/composer.json (hunk 4 above)
Comment (carusogabriel) [reply src/Php73/composer.json]: @stof I'll just wait for the official merge in php-src, I'll do it :)

File: tests/Php73/Php72Test.php (hunk 3 above)
Comment (keradus) [reply tests/Php73/Php72Test.php]: even if not, it shall still be tested

**Comment by carusogabriel** [comment]
@nicolas-grekas Should be ready to go :smiley:

File: tests/Php73/Php73Test.php [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    {
+        $this->assertFalse(is_countable($this->genOneToTen()));
+    }
+
+    private function genOneToTen()
+    {
+        for ($i = 1; $i <= 10; ++$i) {
+            yield $i;
+        }
+    }
```
Comment (carusogabriel) [inline tests/Php73/Php73Test.php]: @keradus Is there a way to add generators tests without breaking the build?

File: tests/Php73/Php73Test.php [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+ * @author Gabriel Caruso <carusogabriel34@gmail.com>
+ */
+class Php73Test extends TestCase
+{
+    /**
+     * @covers Symfony\Polyfill\Php73\Php73::is_countable
+     */
+    public function testIsCountable()
+    {
+        $this->assertTrue(is_countable(array(1, 2, '3')));
```
Comment (carusogabriel) [inline tests/Php73/Php73Test.php]: @keradus Fixed the array syntax, thanks.

File: tests/Php73/Php73Test.php (hunk 5 above)
Comment (keradus) [reply tests/Php73/Php73Test.php]: the `yield` code would need to be extracted to dedicated file and loaded conditionally

Comment (carusogabriel) [reply tests/Php73/Php73Test.php]: @keradus Done, thanks.

**Comment by keradus** [comment]
looks better with every change @carusogabriel !
(btw, if you would not squash, it would be way easier to review the changes, now one need to read whole diff again and again, instead of just patch - don't worry, squash can be done during merging!)

one last thing - could you extend `.travis.yml` matrix to contain 7.2 and 7.3 ?

**Comment by carusogabriel** [comment]
@keradus 
> btw, if you would not squash, it would be way easier to review the changes, now one need to read whole diff again and again, instead of just patch - don't worry, squash can be done during merging!

Sure, thanks for the tip, I always amend my commits, but I'll start to commit and then squash them before merging the PR, thanks!

> could you extend .travis.yml matrix to contain 7.2 and 7.3 ?

Sure, pushing this changes ASAP.

**Comment by carusogabriel** [comment]
@keradus There's one remain error in PHP 7.3 that I couldn't figure out how to solve. Maybe we should just add 7.2?

**Comment by keradus** [comment]
the feature you are adding has to be tested under 7.3. if not, then we test only "does our polyfill work like we expect", but don't "does our polyfill expectation are same as original implementation".
yet, I agree 7.3 failing is not related to your PR.
sth nice to have for sure, but maybe indeed out of the scope here. nevertheless, adding it (even temporary) to travis matrix has proven that your changes works same as native implementation.

if you ask me, I would be fine with reverting travis changes and do it as separated PR.

**Comment by keradus** [comment]
@stof , your request has been already fulfilled, could you please revisit your review ?

In general, missing thing on Travis is related to `Mbstring`, thus it's not connected with feature introduced by this PR.
Could we make this one released?

File: composer.json [hunk 7]
```diff
@@ -50,6 +51,7 @@
             "src/Php70/bootstrap.php",
             "src/Php71/bootstrap.php",
             "src/Php72/bootstrap.php",
+            "src/Php73/bootstrap.php",
```
Comment (nicolas-grekas) [inline composer.json]: the branch-alias should be updated to 1.8-dev for all composer.json files of the repo

File: src/Php73/LICENSE [hunk 8]
```diff
@@ -0,0 +1,19 @@
+Copyright (c) 2015-2018 Fabien Potencier
```
Comment (nicolas-grekas) [inline src/Php73/LICENSE]: 2018 only (not "2015-")

File: src/Php73/bootstrap.php [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+if (PHP_VERSION_ID < 70300) {
+    if (!function_exists('is_countable')) {
+        function is_countable($s) {
```
Comment (nicolas-grekas) [inline src/Php73/bootstrap.php]: the function should be on one single line (to make the test suite work as expected)
`$s` should be renamed by `$var` I suppose also.

**Comment by nicolas-grekas** [comment]
Would you mind adding `hrtime()` also in the PHP 7.3 polyfill (based on `microtime()`)?
```
  . Added monotonic timer function hrtime([bool get_as_num]). It returns an
    array of the form [seconds, nanoseconds] with the timestamp starting at
    an unspecified point in the past. If the optional argument is passed as
    true, the return value is an integer on 64-bit systems or float on
    32-bit systems, representing the nanoseconds. The timestamp is not
    adjustable and is not related to wall clock or time of day. The timers
    are available under Linux, FreeBSD, Windows, Mac, SunOS, AIX and their
    derivatives. If no required timers are provided by a corresponding
    platform, the function returns false.
```

**Comment by keradus** [comment]
Could we please add different feature with different PR?
This one is already awaiting for 2.5 months, no need to postpone it even further ;)

**Comment by nicolas-grekas** [comment]
Thank you @carusogabriel.

**Comment by nicolas-grekas** [comment]
FYI, I removed the nightly line because it failed for unrelated reason, to be fixed in a next PR. hrtime() welcome now.

**Comment by Majkl578** [comment]
As noted in the commit here: https://github.com/symfony/polyfill-php73/commit/f7139426b2a7fce4d4440fde55284fe47a1b18ca#r28736995
This implementation is incorrect.

**Pull request merged**